wgpu-hal/src/dx12/shader_compilation.rs

```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
-mod dxc {
-    use std::ffi::CStr;
-    use std::path::PathBuf;
-
-    // Destructor order should be fine since _dxil and _dxc don't rely on each other.
-    pub(crate) struct DxcContainer {
-        compiler: hassle_rs::DxcCompiler,
-        library: hassle_rs::DxcLibrary,
-        validator: hassle_rs::DxcValidator,
-        // Has to be held onto for the lifetime of the device otherwise shaders will fail to compile.
-        _dxc: hassle_rs::Dxc,
-        // Also Has to be held onto for the lifetime of the device otherwise shaders will fail to validate.
-        _dxil: hassle_rs::Dxil,
-    }
+trait DxcObj: Interface {
+    const CLSID: windows::core::GUID;
+}
+impl DxcObj for Dxc::IDxcCompiler3 {
+    const CLSID: windows::core::GUID = Dxc::CLSID_DxcCompiler;
+}
+impl DxcObj for Dxc::IDxcUtils {
+    const CLSID: windows::core::GUID = Dxc::CLSID_DxcUtils;
+}
+impl DxcObj for Dxc::IDxcValidator {
+    const CLSID: windows::core::GUID = Dxc::CLSID_DxcValidator;
+}
 
-    pub(crate) fn get_dxc_container(
-        dxc_path: Option<PathBuf>,
-        dxil_path: Option<PathBuf>,
-    ) -> Result<Option<DxcContainer>, crate::DeviceError> {
-        // Make sure that dxil.dll exists.
-        let dxil = match hassle_rs::Dxil::new(dxil_path) {
-            Ok(dxil) => dxil,
-            Err(e) => {
-                log::warn!("Failed to load dxil.dll. Defaulting to FXC instead: {}", e);
-                return Ok(None);
-            }
-        };
+#[derive(Debug)]
+struct DxcLib {
+    lib: crate::dx12::DynLib,
+}
 
-        // Needed for explicit validation.
-        let validator = dxil.create_validator()?;
-
-        let dxc = match hassle_rs::Dxc::new(dxc_path) {
-            Ok(dxc) => dxc,
-            Err(e) => {
-                log::warn!(
-                    "Failed to load dxcompiler.dll. Defaulting to FXC instead: {}",
-                    e
-                );
-                return Ok(None);
+impl DxcLib {
+    fn new(lib_path: Option<PathBuf>, lib_name: &'static str) -> Result<Self, libloading::Error> {
+        let lib_path = if let Some(lib_path) = lib_path {
+            if lib_path.is_file() {
+                lib_path
+            } else {
+                lib_path.join(lib_name)
             }
+        } else {
+            PathBuf::from(lib_name)
         };
-        let compiler = dxc.create_compiler()?;
-        let library = dxc.create_library()?;
-
-        Ok(Some(DxcContainer {
-            _dxc: dxc,
-            compiler,
-            library,
-            _dxil: dxil,
-            validator,
-        }))
+        unsafe { crate::dx12::DynLib::new(lib_path).map(|lib| Self { lib }) }
+    }
+
+    pub fn create_instance<T: DxcObj>(&self) -> Result<T, crate::DeviceError> {
+        type Fun = extern "system" fn(
+            rclsid: *const windows_core::GUID,
+            riid: *const windows_core::GUID,
+            ppv: *mut *mut core::ffi::c_void,
+        ) -> windows_core::HRESULT;
```

Comment:
Any reason to not use the existing `DxcCreateInstanceProc` typedef? Because it's wrapped in `Option<>`?

Reply:
I saw that it exists but wasn't 100% sure how that would behave with libloading. It's not something I've seen in their docs/examples.